### PR TITLE
Refactored logout handling to use the sessionIndex to find the session

### PR DESCRIPTION
**Pull request by mcastany**
We were using only the audience to search for the session participant. We should use the NameID and sessionIndex. Update the sessionParticipants get and remove methods to receive the nameId and sessionIndex as parameters.

I've also stored audience, sessionIndex and nameId in the state before issuing a LogoutRequest, so when the flow resumes back, we have all the information in the context to retrieve the session participant

### CODE REVIEW

File: lib/logout.js [hunk 1]
```diff
@@ -209,9 +206,9 @@ module.exports.logout = function (options) {
 
       // Logout flow in progress, incoming SAMLResponse from SP. (Could be SP initiated or IdP initiated)
       } else if (req.query.SAMLResponse || req.body.SAMLResponse) {
-        function processLogoutResponse(logoutResponse, transactionId) {
+        function processLogoutResponse(state, transactionId) {
```
Comment (ziluvatar) [inline lib/logout.js]: Shouldn't we check here that the logout response info is the same as we have in the stored state? at least the issuer, it's weird we ignore the response and only use the info from our state

Comment (mcastany) [inline lib/logout.js]: The logout response only says if the logout succeeded or not. It does not contain a sessionIndex or a nameid. We check that the response was for an existing request (by checking the RelayState), the signature (if signed and configured) and the status.

File: lib/sessionParticipants/index.js [hunk 2]
```diff
@@ -9,13 +9,16 @@ function SessionParticipants (sessions) {
...
+ * @sessionIndex {string}   The string as it was received in the SAML request/response. Only available in LogoutRequests
  * @cb     {function} The callback that will be called with '(err, sessionParticipant)'
  */
-SessionParticipants.prototype.get = function(issuer, cb) {
+SessionParticipants.prototype.get = function(issuer, sessionIndex, nameId, cb) {
+  // SessinIndex should be mandatory, but not issuer
+  // Let's keep using issuer only if available
   var s = this._participants.find(function (session) {
-    return session.serviceProviderId === issuer;
+    return session.serviceProviderId === issuer  && session.sessionIndex === sessionIndex  && session.nameId === nameId;
```
Comment (mcastany) [inline lib/sessionParticipants/index.js]: They shouldn't, as nameid is mandatory. If the three of them are undefined, it won't find any session, throwing an error

File: lib/logout.js [hunk 3]
```diff
@@ -24,12 +24,13 @@ module.exports.logout = function (options) {
     // Finished if there are no more session - finish logout
     if (!sessionParticipants.hasElements()) { return finalize(transactionId, req, res, next); }
 
-    var logoutRequestState = { transactionId: transactionId };
+    sessionParticipants.getFirst(function (err, participant) {
+      // Store the sssionIndex in session so when the logout request comes back we can use that to validate
```
Comment (mcastany) [inline lib/logout.js]: done

Comment (ziluvatar) [inline lib/logout.js]: Did you push it?

Comment (mcastany) [inline lib/logout.js]: Done, sorry fixed it on another file. Now yes :)

File: lib/logout.js (hunk 1 above)
Comment (ziluvatar) [inline lib/logout.js]: Then can you change the name of the function? It's doing nothing with logout response then

Comment (mcastany) [inline lib/logout.js]: ack

Comment (mcastany) [inline lib/logout.js]: done

**Pull request merged**